Adding configuration such that CMake find_package can also find the installed package.

stefi07 opened the pull request:
Export library to ImageStreamIOTargets.cmake, which gets installed in lib/cmake (following pkgconfig).
Generate ImageStreamIOConfigVersion.cmake and ImageStreamIOConfig.cmake, which are also installed in lib/cmake.

stefi07 committed:
Adding configuration such that CMake find_package can also find the
installed package. Following the same procedure as for milk.

jaredmales commented:
@oguyon @DasVinch any issues with this?

DasVinch commented on CMakeLists.txt:
```diff
@@ -9,7 +9,7 @@ message(" SRCNAME = ${SRCNAME} -> LIBNAME = ${LIBNAME}")
 option(gtest_build_tests "Build Unit Tests" OFF)
 
-project(${LIBNAME} LANGUAGES C)
+project(${LIBNAME} VERSION 1.0 LANGUAGES C)
```
I believe we're at 1.4
Will this be overwritten from what ImageStreamIO.h says, or will this value be the decisive version number?

DasVinch replied:
Edit: we're at 2.0, and ImageStruct.h is the file that says so.

stefi07 replied:
Ah, I missed that a different version was specified, thanks for pointing that out. I've updated the CMakeLists.txt to read the version from the ImageStruct.h macro.

stefi07 replied:
Edit: replaced that with a system similar to milk.

DasVinch commented:
No beef here!
I was just making sure I understood the why on the sibling milk PR. Looks good to me! Thanks for contributing.

Actually there's just this tiny tweak about the version number -- see the review I made.
I think for milk the version number is set somewhere, then exported to the dynamically generated milk_config.h by cmake at compile time. Maybe we should just make the same strategy happen here in ImageStreamIO?

stefi07 committed:
Now reading version from ImageStruct.h macro definition

stefi07 committed:
Updated version config to mirror that of milk

Created ImageStreamIO_config.h.in that defines project name & version
params.
Version gets set in CMakeLists.txt and config header gets dynamically
generated.
Updated ImageStruct.h to include ImageStreamIO_config.h.
Added ImageStreamIO_config.h to .gitignore.

stefi07 commented:
Yup, that makes sense. I've made updates such that a ImageStreamIO_config.h gets dynamically generated during the build and that is read in by ImageStruct.h. This makes CMakeLists.txt the source of truth for the version number (I've added more details on the specific changes in the [commit message](https://github.com/milk-org/ImageStreamIO/pull/64/commits/df6554720bba14c9038befbda1852c24ad3d6139)).

DasVinch commented:
Looks good to me! Thanks again.

I'll make a notional comment on how we increment ISIO versions at the package vs the ImageStruct level.
Probably we should opt for minor increments for API changes and major increments for struct changes, and change the compatibility requirements accordingly (although, almost no one ever would end up running to ISIO instances with different versions on the same system). @oguyon thoughts?

DasVinch commented:
Uh Oh I made a mistake...

Basically using the same strategy makes milk's config.h and ISIO's config.h use #defines with identical names...

This makes the #include system fairly unhappy.

The answer is available here:
https://stackoverflow.com/questions/7398/how-to-avoid-redefining-version-package-etc

> It is usually a big mistake to have config.h be included from public headers. config.h is always private to your project or the subproject, and should only be included from .c files. So, if your vendor's documentation says to include their "vendor.h" and that public header includes config.h somehow, then that is a no-no. Similarly, if your project is a library, don't include config.h anywhere from your publically installed headers.

So I'm gonna put an emergency commit to fix that and merge your PR + hotfix it in milk as well.
